src/view.rs

```diff
@@ -15,23 +36,151 @@ impl View {
+            State::Lines => self.state = State::Clean,
+            _ => {}
+        }
+    }
+
+    fn set_clean_cursor(&mut self) {
+        match self.state {
+            State::All => self.state = State::Lines,
+            State::Cursor => self.state = State::Clean,
+            _ => {}
+        }
+    }
+
+    pub fn render<W: Write>(&mut self, w: &mut W, height: u16) -> Result<()> {
+        if self.dirty_lines() {
+            write!(w, "{}{}", cursor::Goto(1, 1), clear::All)
+                .chain_err(|| ErrorKind::DisplayError)?;
+
+            self.render_lines(w, height)?;
+            self.set_clean_lines();
+        }
+
+        if self.dirty_cursor() {
+            self.render_cursor(w)?;
+            self.set_clean_cursor();
+        }
+        Ok(())
+    }
+
+    fn render_lines<W: Write>(&self, w: &mut W, height: u16) -> Result<()> {
+        let mut invalid_lines: usize = 0;
+        for (lineno, line) in self.lines.iter().enumerate() {
+            if line.is_valid {
+                // Lines are drawn with fake cursors.
+                // We set the actual cursor later, redrawing the line in the process.
+                line.render(w, (lineno - invalid_lines) as u16 + 1, None)?;
+            } else {
+                invalid_lines += 1;
+            }
+            if lineno > invalid_lines && (lineno - invalid_lines) == height as usize {
+                break;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn render_cursor<W: Write>(&self, w: &mut W) -> Result<()> {
+        if let Some(cursor) = self.cursor.as_ref() {
+            if cursor.line as usize <= self.lines.len() {
+                // Redraw the line without the fake cursor
+                let line = self.lines
+                    .get(cursor.line as usize)
+                    .and_then(|line| if line.is_valid { Some(line) } else { None })
+                    .ok_or_else(|| {
+                        error!("No valid line at cursor index {}", cursor.line);
+                        ErrorKind::DisplayError
+                    })?;
+                line.render(w, cursor.line + 1, Some(cursor))?;
+
+                // Draw the cursor. The trick is that some characters are larger than others.
+                //
+                // For the moment, we only handle tabs, and we assume the terminal has tabstops of
+                // TAB_LENGTH.
+                let column = line.text
+                    .as_ref()
+                    .map(|s| &**s)
+                    .unwrap_or("")
+                    .chars()
+                    .take(cursor.column as usize)
+                    .fold(0, |acc, c| {
+                        if c == '\t' {
+                            acc + TAB_LENGTH - (acc % TAB_LENGTH)
+                        } else {
+                            acc + 1
+                        }
+                    });
+                let cursor_pos = cursor::Goto(column as u16 + 1, cursor.line + 1);
```

Comment:
let's cast `column` above, so this call looks nicer

(also, if you move the computation to its own method, the return type can be u16 instead)